core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCache.java

```diff
@@ -402,19 +426,37 @@ public T retry() {
          * a special case that looks for Code.NONODE in the KeeperException, then non-existence can
          * not be cached.
          */
+
+        Matcher configMatcher = TABLE_SETTING_CONFIG_PATTERN.matcher(zPath);
+        boolean createWatch = !configMatcher.matches();
+
         cacheWriteLock.lock();
         try {
           final ZooKeeper zooKeeper = getZooKeeper();
-          Stat stat = zooKeeper.exists(zPath, watcher);
+          Stat stat = zooKeeper.exists(zPath, createWatch ? watcher : null);
           byte[] data = null;
           if (stat == null) {
             if (log.isTraceEnabled()) {
               log.trace("zookeeper did not contain {}", zPath);
             }
           } else {
             try {
-              data = zooKeeper.getData(zPath, watcher, stat);
+              data = zooKeeper.getData(zPath, createWatch ? watcher : null, stat);
               zstat = new ZcStat(stat);
+              // Create the first watcher for /accumulo/{InstanceID/table-config-version
+              // so that the next modification of a table config in ZooUtil.putData
+              // will trigger a watch event for that  znode.
+              // Watchers need to be reset on a node if a watch event was triggered.
+              // The reset of the watcher for the above mentioned node will be done in the NodeDataChanged
+              // condition of the ZCacheWatcher.process function when it has been triggered
+              // by a change created in ZooUtil.putData.
+              if (!tableConfigWatcherSet) {
+                if (configMatcher.matches()) {
+                  String pathPrefix = configMatcher.group(1);
+                  zooKeeper.exists(pathPrefix + Constants.ZTABLE_CONFIG_VERSION, watcher);
```

Comment:
It can't go into the watcher process event or that event will never be called because the "table-config-version" will not have a "watcher" on it. 

I think one of the ZooCache constructors is a create place to "put the watch" on "table-config-version" but I don't have a way to find out the accumulo instance ID there.  That is why I have to "regex" it out of one of the table config zpaths when I get  chance to.  

I need to find a better way to put on the initial watch.  Agreed.

Reply:
Now I am really confused.  I thought we were putting a watcher on the table-config-version and only that node instead of watching the nodes under the table/conf directories?

Reply:
OK.  I can make this modification to the code so we can get passed this issue.  I will do it tomorrow.